src/System.Runtime.Analyzers/UnitTests/TestForNaNCorrectlyTests.cs

```diff
@@ -1,23 +1,409 @@
+        {
+            var code = @"
+Public Class A
+    Public Function Compare(f As Single) As Boolean
+        Return f = Single.NbN   ' Misspelled
+    End Function
+End Class
+";
+            VerifyBasic(code);
+        }
+
+        [Fact]
+        public void CSharpNoDiagnosticForComparisonWithFunctionReturningNaN()
+        {
+            var code = @"
+public class A
+{
+    public bool Compare(float f)
+    {
+        return f == NaNFunc();
+    }
+
+    private float NaNFunc()
+    {
+        return float.NaN;
+    }
+}
+";
+            VerifyCSharp(code);
+        }
+
+        [Fact]
+        public void BasicNoDiagnosticForComparisonWithFunctionReturningNaN()
+        {
+            var code = @"
+Public Class A
+    Public Function Compare(f As Single) As Boolean
+        Return f = NaNFunc()
+    End Function
+
+    Private Function NaNFunc() As Single
+        Return Single.NaN
+    End Function
+End Class
+";
+            VerifyBasic(code);
+        }
+
+        [Fact]
+        public void CSharpNoDiagnosticForEqualityWithNonNaN()
+        {
+            var code = @"
+public class A
+{
+    public bool Compare(float f)
+    {
+        return f == 1.0;
+    }
+}
+";
+            VerifyCSharp(code);
+        }
+
+        [Fact]
+        public void BasicNoDiagnosticForEqualityWithNonNaN()
+        {
+            var code = @"
+Public Class A
+    Public Function Compare(f As Single) As Boolean
+        Return f = 1.0
+    End Function
+End Class
+";
+            VerifyBasic(code);
+        }
+
+        [Fact]
+        public void CSharpNoDiagnosticForNonComparisonOperationWithNaN()
+        {
+            var code = @"
+public class A
+{
+    public float OperateOn(float f)
+    {
+        return f + float.NaN;
+    }
+}
+";
+            VerifyCSharp(code);
+        }
+
+        [Fact]
+        public void BasicNoDiagnosticForNonComparisonOperationWithNonNaN()
+        {
+            var code = @"
+Public Class A
+    Public Function OperateOn(f As Single) As Single
+        Return f + Single.NaN
+    End Function
+End Class
+";
+            VerifyBasic(code);
+        }
+
+        [Fact]
+        public void CSharpOnlyOneDiagnosticForComparisonWithNaNOnBothSides()
+        {
+            var code = @"
+public class A
+{
+    public bool Compare()
+    {
+        return float.NaN == float.NaN;
+    }
+}
+";
+            VerifyCSharp(code, GetCSharpResultAt(6, 16));
+        }
+
+        [Fact]
+        public void BasicOnlyOneDiagnosticForComparisonWithNonNaNOnBothSides()
+        {
+            var code = @"
+Public Class A
+    Public Function Compare() As Boolean
+        Return Single.NaN = Single.NaN
+    End Function
+End Class
+";
+            VerifyBasic(code, GetBasicResultAt(4, 16));
+        }
+
+        // TODO: Not really testing bad expression right.
+        // TODO: Other kinds of syntax errors:
+        //  Wrong comparison operator
+        //  Wrong data type (float <=> Single)
```

Comment:
It might be interesting to test the comparisons in other places in code : method arguments, ternary expressions etc. to make sure operation tree is correct in those situations.